### DISCUSSION TITLE
Control Window Order when Autotiling

### DISCUSSION THREAD

**Discussion by adhadse** [gTile/gTile#396] [Ideas]
I've configured gTile's Autotiling to have `2,3`/`3,2`.

When I try to use Autotiling feature it prefers a specific window to the right side of the monitor. I wanted the *main* selected window to be on the right side. 

So even if the window I *snapped* to the right side switches it's position to left side instead of remaining on right side and resizing itself to configured Autotiling size.

The PyCharm winodw snapped to right switches it's position to left. Can we avoid doing that with Autotiling or should I do something else? Any suggestions?

[Screencast from 2024-02-04 10-47-56.webm](https://github.com/gTile/gTile/assets/56764399/e029d036-2a3f-4492-a9f7-c1a339efa915)

**Comment by schnz** [comment]
Unfortunately, there is no particular feature which you can use to specify the placement of windows. Windows are placed in the order that the Gnome Shell API returns them. From what I can tell, the only way to influence the window placement location as of now, is to open the windows in reverse order to how you want them to be placed  (i.e. from right to left).

If you have a good idea/suggestion on how the preferred window could be specified, I would add this as a feature to the backlog.

**Comment by scherepanov** [comment]
@schnz 
This user inquery is a very good example what can be improved in gTile.
We need internal API for tiling. 
I am thinking of having base tiling class implementing basic services, and having derived class implementing tiling function.
Tiling function will get on input list of windows as returned by Gnome API with their positions. It will return list of windows that need to change their positions/size, along with their new positions/sizes. (It will be more input data, but on most basic level just list of windows with their positions/sizes would do).
Base tiling class will provide integration with shortcuts and preferences dialog etc. We should be able to include new derived tiling settings into preferences automatically.

This specific user request do need some coding. I am looking to have minimal coding to add new tiling request into gTile.

Ideally, we can add code for new derived class in preferences itself, making it completely dynamic. New code willb e executed in JS sandbox, and will have access only to data we will give to input parameters to function, and will not be able to change anything else except providing output data - list of windows to resize, along with new sizes/positions.

If we would have this functionality, implementing user request will take around 10 min and do not need any code changes. Create new tiling in preferences, add custom code to tiling function, link to keyboard shortcuts, and we are done.

**Comment by scherepanov** [comment]
@schnz 
Dynamic code execution do represent a security risk.
Quick internet check show that JS sandbox has not been broken (yet). Tiling function with dynamic code execution from preferences dialog  is secure.

**Comment by schnz** [comment]
**Off-Topic**

Sorry that I picked this issue as means for communication. 

@scherepanov: There are quite a few issues which are more or less feature-requests. I propose to convert all of these issues into a [Discussion](https://github.com/gTile/gTile/discussions/categories/ideas). As it stands right now there is not much activity in terms of active development from other developers and I would find it helpful if the issue tracker is used for issues that require active attention (i.e. bugs, regressions, etc.).

I don't want to close issues like this one because they have a legitimate reason to exist. But they are long standing issues that are no longer need the active attention from us maintainers. And unless someone is willing to implement these requests feature-wise, they are also not going to be closed any time soon.

What do you think?

**Comment by scherepanov** [comment]
Hi @schnz , sorry for delay with answer, I am on vacations this week.
Absolutely no problems to convert feature requests into discussions. Please go ahead.

**Comment by sbroenner** [comment]
Hello everyone, sorry to dig up this old thread but I think my problem is probably related. I've just upgrade my Debian install from Bookworm to Trixie, which has apparently triggered an update of gTile due to the change from Gnome 43.x to 48.x. The version of gTile I have installed is v62.

Unfortunately autotile is now acting completely differently - the active window no longer tiles to the top leftmost column/tile, but now randomly either ends up on the right, center or bottom left, depending on how many other windows are open.

Is this expected behavior that ties into this reply?

> Unfortunately, there is no particular feature which you can use to specify the placement of windows. Windows are placed in the order that the Gnome Shell API returns them. From what I can tell, the only way to influence the window placement location as of now, is to open the windows in reverse order to how you want them to be placed (i.e. from right to left).

This doesn't really explain the apparent randomness I'm seeing.

If this is not expected behavior, please give me a heads up and I will try to figure out if it's another extension mucking things up, or maybe my settings from the old version of gTile are doing something weird...

-edit- Just reset all gTile settings to default - didn't change anything about the behaviour of autotile.

**Reply by schnz** [reply]
Hi @sbroenner,

random behavior is _not_ expected! Do you remember how the windows were ordered in the legacy version of gTile? If there was a deterministic behavior, we may be able to find a solution and go back to that behavior.

**Reply by sbroenner** [reply]
Hi @schnz ,

luckily I still have a machine or two running Bookworm, so checking how the older versions behaved is not an issue.

The previous non-random behavior **when using the autotile function via a number key** was as follows:

* The active window was always tiled to the very top left corner (or just the left if there was only a single row)
* Other non-focused windows were then added from left to right - the order of these, as far as I could tell, was the order they had been opened in
* Selecting a different window as the active window and triggering auto-tile (same number of columns as before) would move that window to the top left and re-tile all other windows beside it
* The layout was always consistent, e.g. the windows were all the same size regardless of gaps that might result (e.g. 5 windows autotiled to 3 columns)
* Gaps were always on the bottom right due to the tiles being filled up from the top left

In the new behavior:

* The area occupied by the focused window depends on the number of open windows and the number of columns chosen
* The "randomness" I mentioned refers mainly to where the active window ends up in the grid - It's not truly random, I've just not worked out why it ends up where it ends up yet.
* 6 windows in 3 columns seems to favor the center column for the active window - sometimes top middle or bottom middle
* 6 windows in 4 columns seems to favor the second column from the left
* 2 windows in 2 columns seems to favor the right column instead of the left (previous behaviour)
* Left over gaps are filled dynamically with adjacent windows, but without being able to determine which ones end up there, this feels like random windows are bigger than others


It's possible I may simply be missing some user instructions or an explanation of the new behaviour that will make it appear less random to me. My main objective here is to find out whether the new behaviour I'm describing is even normal...

**Reply by schnz** [reply]
Thanks for the detailed explanation! I probably also have to setup a VM to check out the old behavior of gTile. From what I understand, the autotile feature that you are used to, was automatically determining the amount of columns, based on the number of windows on the screen? (e.g., 5 windows resulted in 3 columns)? This is indeed different from the new behavior but could be reverted. Just to be clear: We are talking about the autotiling feature that (by default) uses the `m` keyboard shortcut?

**Reply by sbroenner** [reply]
>From what I understand, the autotile feature that you are used to, was automatically determining the amount of columns, based on the number of windows on the screen? (e.g., 5 windows resulted in 3 columns)?

Not quite - I'm specifically talking about the autotile function that is triggered **with a number key**, e.g. you press Super+Enter to open gTile and then press a number key corresponding to the number of columns you want - e.g. you want 3 columns, press 3. Want 5 columns, press 5.

In the example with 5 windows, I would have actively chosen to have 3 columns, resulting in all 5 windows tiled to a 3x2 grid with a gap in the bottom right.


The "m" auto tiling appears to more or less still work as before - the active window is put on the left and the others are tiled horizontally to the right of it. This is not my preferred workflow so I don't use it a lot, but it does appear to still work.

**Reply by schnz** [reply]
Understood. This logic was reworked quite a bit, which explains some of the points you mentioned:


> The layout was always consistent, e.g. the windows were all the same size regardless of gaps that might result (e.g. 5 windows autotiled to 3 columns
> Gaps were always on the bottom right due to the tiles being filled up from the top left

This was indeed an intentional regression. IIRC, there were situations in the legacy version of gTile, where there was either (i) empty/unused spaced after autotiling or (ii) some windows were overlayed by other windows after autotiling. The intention was to always display _all_ windows while using _all_ of the available space.

---

However, the "random" order of the windows was not intended and is something I'll have a look at.

**Reply by sbroenner** [reply]
>However, the "random" order of the windows was not intended and is something I'll have a look at.

Excellent, thank you. If you need any more information/screenshots/videos, especially for the older versions, I'm happy to help.

>This was indeed an intentional regression.
>...
>>The intention was to always display all windows while using all of the available space.

I wouldn't even call this part a regression - using all the available space does make sense to me, of course. It just triggers user irritation because naturally we want to be able to determine which of the open windows receives the privilege of that extra space ;).
